lib/shortcode/parser.rb

```diff
@@ -1,36 +1,38 @@
 
-  rule(:space)        { str(' ').repeat(1) }
-  rule(:space?)       { space.maybe }
-  rule(:newline)      { (str("\r\n") | str("\n")) >> space? }
-  rule(:newline?)     { newline.maybe }
-  rule(:whitespace)   { (space | newline).repeat(1) }
+    self.class.rule(:space)        { str(' ').repeat(1) }
+    self.class.rule(:space?)       { space.maybe }
+    self.class.rule(:newline)      { (str("\r\n") | str("\n")) >> space? }
+    self.class.rule(:newline?)     { newline.maybe }
+    self.class.rule(:whitespace)   { (space | newline).repeat(1) }
 
-  rule(:key) { match('[a-zA-Z0-9\-_]').repeat(1) }
+    self.class.rule(:key) { match('[a-zA-Z0-9\-_]').repeat(1) }
 
-  rule(:value_with_quotes) { quotes >> (quotes.absent? >> any).repeat.as(:value) >> quotes }
-  rule(:value_without_quotes) { quotes.absent? >> ( (str(']') | whitespace).absent? >> any ).repeat.as(:value) }
-  rule(:value) { Shortcode.configuration.use_attribute_quotes ? value_with_quotes : (value_without_quotes | value_with_quotes) }
+    self.class.rule(:value_with_quotes) { quotes >> (quotes.absent? >> any).repeat.as(:value) >> quotes }
+    self.class.rule(:value_without_quotes) { quotes.absent? >> ( (str(']') | whitespace).absent? >> any ).repeat.as(:value) }
+    self.class.rule(:value) { configuration.use_attribute_quotes ? value_with_quotes : (value_without_quotes | value_with_quotes) }
 
-  rule(:option)   { key.as(:key) >> str('=') >> value }
-  rule(:options)  { (str(' ') >> option).repeat(1) }
-  rule(:options?) { options.repeat(0, 1) }
+    self.class.rule(:option)   { key.as(:key) >> str('=') >> value }
+    self.class.rule(:options)  { (str(' ') >> option).repeat(1) }
+    self.class.rule(:options?) { options.repeat(0, 1) }
 
-  rule(:open)       { str('[') >> block_tag.as(:open) >> options?.as(:options) >> str(']') >> newline? }
-  rule(:close)      { str('[/') >> block_tag.as(:close) >> str(']') >> newline? }
-  rule(:open_close) { str('[') >> self_closing_tag.as(:open_close) >> options?.as(:options) >> str(']') >> newline? }
+    self.class.rule(:open)       { str('[') >> block_tag.as(:open) >> options?.as(:options) >> str(']') >> newline? }
+    self.class.rule(:close)      { str('[/') >> block_tag.as(:close) >> str(']') >> newline? }
+    self.class.rule(:open_close) { str('[') >> self_closing_tag.as(:open_close) >> options?.as(:options) >> str(']') >> newline? }
 
-  rule(:text)   { ((close | block | open_close).absent? >> any).repeat(1).as(:text) }
-  rule(:block)  { (open >> (block | text | open_close).repeat.as(:inner) >> close) }
-
-  rule(:body) { (block | text | open_close).repeat.as(:body) }
-  root(:body)
+    self.class.rule(:text)   { ((close | block | open_close).absent? >> any).repeat(1).as(:text) }
+    self.class.rule(:block)  { (open >> (block | text | open_close).repeat.as(:inner) >> close) }
 
+    self.class.rule(:body) { (block | text | open_close).repeat.as(:body) }
+    self.class.root(:body)
+  end
```

Comment:
All the rules for the parser are evaluated in the constructor, based on the injected configuration.